### DISCUSSION TITLE
FacetGrid map_dataarray usage?

### DISCUSSION THREAD

**Discussion by joshdorrington** [pydata/xarray#7310] [Q&A] [answered]
I am trying to use FacetGrid to plot a number of spatial fields, and then overlay significance contours on them.
I have a multidimensional Dataarray like this:

<xarray.DataArray 'Z500' (mode: 2, region: 7, index_val: 2, lat: 91, lon: 360)>
[917280 values with dtype=float32]
Coordinates:
  * region     (region) object 'NE' 'NW' 'SE' 'SW' 'SWNW' 'SWSE' 'SWSENW'
  * lon        (lon) float32 -180.0 -179.0 -178.0 -177.0 ... 177.0 178.0 179.0
  * lat        (lat) float32 0.0 1.0 2.0 3.0 4.0 ... 86.0 87.0 88.0 89.0 90.0
  * index_val  (index_val) int64 1 2
  * mode       (mode) object 'C' 'S'

Where I am storing the field values in mode='C', and the boolean significance test in mode='S'. I tried the following approach:

```
p=da.sel(mode='S').plot.contour(col='region',row='index_val') #Plot my significance levels
p.data=da.sel(mode='C') #Change the data
p.map_dataarray(xr.plot.pcolormesh, y="region", x="index_val") #Plot my field

```
However this fails on the third line with the error:
`IndexError: tuple index out of range`

This is a result of the following attempt to access a 'dims' attribut which is just an empty list:
```
File ~/miniconda3/lib/python3.9/site-packages/xarray/plot/plot.py:1147, in _plot2d.<locals>.newplotfunc(darray, x, y, figsize, size, aspect, ax, row, col, col_wrap, xincrease, yincrease, add_colorbar, add_labels, vmin, vmax, cmap, center, robust, extend, levels, infer_intervals, colors, subplot_kws, cbar_ax, cbar_kwargs, xscale, yscale, xticks, yticks, xlim, ylim, norm, **kwargs)
   1145     dims = darray.dims
   1146 else:
-> 1147     dims = (yval.dims[0], xval.dims[0])
```

Can anyone tell me how to make this work?
____________________________________________________________
Full stack trace:
```
IndexError                                Traceback (most recent call last)
Input In [158], in <cell line: 4>()
      1 p=da.sel(mode='S').plot.contour(col='region',row='index_val')
      3 p.data=da.sel(mode='C')
----> 4 p.map_dataarray(xr.plot.pcolormesh, y="region", x="index_val")

File ~/miniconda3/lib/python3.9/site-packages/xarray/plot/facetgrid.py:283, in FacetGrid.map_dataarray(self, func, x, y, **kwargs)
    281     if d is not None:
    282         subset = self.data.loc[d]
--> 283         mappable = func(
    284             subset, x=x, y=y, ax=ax, **func_kwargs, _is_facetgrid=True
    285         )
    286         self._mappables.append(mappable)
    288 self._finalize_grid(x, y)

File ~/miniconda3/lib/python3.9/site-packages/xarray/plot/plot.py:1147, in _plot2d.<locals>.newplotfunc(darray, x, y, figsize, size, aspect, ax, row, col, col_wrap, xincrease, yincrease, add_colorbar, add_labels, vmin, vmax, cmap, center, robust, extend, levels, infer_intervals, colors, subplot_kws, cbar_ax, cbar_kwargs, xscale, yscale, xticks, yticks, xlim, ylim, norm, **kwargs)
   1145     dims = darray.dims
   1146 else:
-> 1147     dims = (yval.dims[0], xval.dims[0])
   1149 # May need to transpose for correct x, y labels
   1150 # xlab may be the name of a coord, we have to check for dim names
   1151 if imshow_rgb:
   1152     # For RGB[A] images, matplotlib requires the color dimension
   1153     # to be last.  In Xarray the order should be unimportant, so
   1154     # we transpose to (y, x, color) to make this work.

IndexError: tuple index out of range
```

**Comment by TomNicholas** [comment]
Hi Josh! 

I'm not sure what is happening here, but it could potentially be a bug, as normally we try to give more informative errors than the one you are seeing. 

> a 'dims' attribute which is just an empty list

This should only be the case for scalar data, and presumably you don't intend to only have one scalar value per pcolormesh plot? I'm not sure if that's because of a problem with your input or our code though.

If you could provide a [minimal reproducible example](https://github.com/pydata/xarray/discussions/5404) (doesn't need to use your real data) it would make it much easier for us to help you debug this.

**Comment by joshdorrington** [comment, marked as answer]
Hi Tom, hope you're well!
I realise now I was making a basic error, and misinterpreting the keyword arguments. For posterity:


```
da=xr.DataArray(data=np.random.randn(2,2,2,10,10),coords={'A':['a1','a2'],'B':[0,1],'C':[0,1],'X':range(10),'Y':range(10)})

#Plot the data on a FacetGrid, column 'B' against row 'C', with contours over 'X' and 'Y'
p=da.sel(A='a1').plot.contour(col='B',row='C') 
try:
    p.map_dataarray(xr.plot.pcolormesh, y="B", x="C") 
except:
    print("keywords 'y' and 'x' are not equivalent to 'col' and 'row'!!")
p.map_dataarray(xr.plot.pcolormesh, y="Y", x="X") 

```
![image](https://user-images.githubusercontent.com/35994700/204387695-a77f3402-86a0-44d7-b94c-2f76f0204e5c.png)

**Comment by joshdorrington** [comment]
There may perhaps be a small bug anyway, as DataArrays with and without coords are handled differently. Contrast:

```
da=xr.DataArray(data=np.random.randn(2,2,2,10,10),coords={'A':['a1','a2'],'B':[0,1],'C':[0,1],'X':range(10),'Y':range(10)})

p=da.sel(A='a1').plot.contour(col='B',row='C') 
try:
    p.map_dataarray(xr.plot.pcolormesh, y="B", x="C");
except Exception as e:
    print('An uninformative error:')
    print(e)
```
```
An uninformative error:
tuple index out of range

```

with:

```
da=xr.DataArray(data=np.random.randn(2,2,2,10,10))

p=da.sel(dim_0=0).plot.contour(col='dim_1',row='dim_2') 
try:
    p.map_dataarray(xr.plot.pcolormesh, y="dim_1", x="dim_2");
except Exception as e:
    print('A more informative error:')
    print(e)
```




```
A more informative error:
x must be one of None, 'dim_3', 'dim_4'

```
